chore: create tests for each API pipeline stage

kylemumma opened the pull request:
Added a general case test for `EntityProcessingStage`, `StorageProcessingStage`, and `ExecutionStage`.

I decided not to test `EntityAndStoragePipelineStage` since it will be phased out soon.

kylemumma commented on tests/pipeline/test_entity_processing_stage.py:
```diff
@@ -0,0 +1,102 @@
+        )
+    )
+    assert res.data and not res.error
+    assert res.query_settings == query_settings and res.timer == timer
```
i did this to make sure they were the same, but in hindsight maybe ill delete this

kylemumma commented on tests/pipeline/test_entity_processing_stage.py:
```diff
@@ -0,0 +1,102 @@
+        and actual.get_granularity() is None
+        and actual.get_limit() == 1000
+    )
+    assert actual.get_from_clause().storage_key == StorageKey.METRICS_DISTRIBUTIONS
```
I did this because I didnt want to recreate the entire Table object, but im not sure if its necessary to make sure it exactly matches or not...

volokluev commented on tests/pipeline/test_storage_processing_stage.py:
```diff
@@ -0,0 +1,88 @@
+
+def test_basic():
+    ch_query = Query(
+        from_clause=Table(
+            "metrics_distributions_v2_local",
+            ColumnSet([]),
+            storage_key=StorageKey.METRICS_DISTRIBUTIONS,
+        ),
```
These tests should not use existing storages/entities, they should be independent of what we already have defined

volokluev commented on tests/pipeline/test_storage_processing_stage.py:
```diff
@@ -0,0 +1,88 @@
+            error=None,
+        )
+    )
+    assert res.data == Query(
```
so as we look at this test, what did it actually test? Have we actually asserted that the pipeline works as we expect it to?

kylemumma replied:
I didn't have a solid understanding of what the storage pipeline is supposed to do, I read through it carefully to understand what it is supposed to do better and modified my test accordingly

volokluev commented on tests/pipeline/test_execution_stage.py:
```diff
@@ -0,0 +1,98 @@
+        ),
+        limit=1000,
+    )
+    res = ExecutionStage(
```
This stage does more than return the query, there are side effects that our application depends on, those should be tested.

volokluev commented on tests/pipeline/test_entity_processing_stage.py:
```diff
@@ -0,0 +1,102 @@
+        and actual.get_granularity() is None
+        and actual.get_limit() == 1000
+    )
+    assert actual.get_from_clause().storage_key == StorageKey.METRICS_DISTRIBUTIONS
```
we should recreate the table object, and again, this should be using a set of entities/storages that are created for this test

kylemumma replied:
done

volokluev reviewed (commented):
Good start, but these tests should be living/enforced documentation of what these stages do, right now, that is not what they are

volokluev commented on tests/pipeline/test_execution_stage.py:
```diff
@@ -0,0 +1,98 @@
+
+@pytest.mark.clickhouse_db
+@pytest.mark.redis_db
+def test_basic():
```
the execution stage has a crucial step of writing to the querylog. We should make sure that happens in this test

volokluev replied:
Also missing a test for dry run queries

volokluev replied:
> it looks like writing to the querylog happens outside of the execution stage

ah okay that's good 😌 

still should add the dry run test though

volokluev commented on tests/pipeline/test_execution_stage.py:
```diff
@@ -0,0 +1,98 @@
+                ]
+            ),
+            storage_key=StorageKey.METRICS_DISTRIBUTIONS,
+        ),
```
Given that this is a high level test, this should include an allocation policy. The execution stage executes allocation policies which is quite important.

volokluev commented on tests/pipeline/test_execution_stage.py:
```diff
@@ -0,0 +1,98 @@
+            "blah",
+            timer,
+        ),
+    ).execute(
```
While this is not a great thing, this is part of the stage at this time, we apply sampling in the execution stage right now, we should make sure that it does happen

https://github.com/getsentry/snuba/blob/14d84ada93e04f3f12f82a8a813d1d029ec627c7/snuba/pipeline/stages/query_execution.py#L292-L310

kylemumma commented:
## major changes
### entity stage
* using new/fake/isolated entity/storage

### storage stage
* using new/fake/isolated entity/storage
* got rid of mocks, now I have a simple query processor that I validate gets run

### execution stage
* added test for dry_run
* added simple allocation policy and make sure its applied
* added test for turbo_sampling

volokluev commented on tests/pipeline/test_entity_processing_stage.py:
```diff
@@ -0,0 +1,127 @@
+    ),
+    readiness_state=ReadinessState.COMPLETE,
+)
+_STORAGE_SET_CLUSTER_MAP[storage.get_storage_set_key()] = CLUSTERS[0]
```
you should put this in a pytest fixture that gets undone after the test completes

volokluev commented on tests/pipeline/test_entity_processing_stage.py:
```diff
@@ -0,0 +1,127 @@
+    required_time_column="timestamp",
+    storage_selector=DefaultQueryStorageSelector(),
+)
+override_entity_map(entkey, entity)
```
here as well

volokluev commented on tests/pipeline/test_entity_processing_stage.py:
```diff
@@ -0,0 +1,127 @@
+entity = PluggableEntity(
+    entity_key=entkey,
+    storages=[EntityStorageConnection(storage, TranslationMappers())],
+    query_processors=[],
```
why no query processors here?

volokluev commented on tests/pipeline/test_execution_stage.py:
```diff
@@ -0,0 +1,190 @@
+    )
+    settings = HTTPQuerySettings()
+    timer = Timer("test")
+    res = ExecutionStage(attinfo, query_metadata=mockmetadata).execute(
```
this query_metadata object gets updated on execution, the variable local to the test and you should probably check what's in it after this is done

volokluev commented on tests/pipeline/test_execution_stage.py:
```diff
@@ -0,0 +1,190 @@
+        and len(res.data.result["data"]) == 1
+        and "avg(duration)" in res.data.result["data"][0]
+    )
+    assert policy.did_update
```
if you're only using the policy in this test, it should be instantiated in this test, right now this test has global state which is a recipe for bugs

volokluev commented on tests/pipeline/test_storage_processing_stage.py:
```diff
@@ -0,0 +1,85 @@
+    allocation_policies=[],
+)
+get_config_built_storages()[mock_storage.get_storage_key()] = mock_storage
+_STORAGE_SET_CLUSTER_MAP[mock_storage.get_storage_set_key()] = CLUSTERS[0]
```
this like this should be done in fixtures

volokluev reviewed (commented):
Good overall, just needs less global state

kylemumma commented:
## major changes
* got rid of global state and replace for fixtures with proper setup & teardown
* added some validation to the metadata object in execution stage test
* added entity processor to entity stage test
